src/psyclone/tests/psyir/nodes/array_mixin_test.py

```diff
@@ -335,47 +369,80 @@ def test_get_lbound_expression_unknown_size(extent):
+def test_get_ubound_expression_unknown_size(extent):
+    '''
+    Test the get_ubound_expression when we have the definition of the
+    array type but its dimensions are unknown.
+
+    '''
+    symbol = DataSymbol("my_symbol", ArrayType(INTEGER_TYPE,
+                                               [extent, extent]))
+    aref = ArrayReference.create(symbol, [_ONE.copy(), _ONE.copy()])
+    ubnd = aref.get_ubound_expression(1)
+    assert isinstance(ubnd, BinaryOperation)
+    assert ubnd.operator == BinaryOperation.Operator.UBOUND
+    assert ubnd.children[0].symbol is symbol
+
+
 def test_aref_to_aos_lbound_expression():
     '''
     Test the get_lbound_expression() method for an ArrayReference to an array
     of structures.
 
     '''
     sgrid_type = StructureType.create(
-        [("id", INTEGER_TYPE, Symbol.Visibility.PUBLIC)])
+        [("ID", INTEGER_TYPE, Symbol.Visibility.PUBLIC)])
     sgrid_type_sym = DataTypeSymbol("subgrid_type", sgrid_type)
     sym = DataSymbol("subgrids", ArrayType(sgrid_type_sym, [(3, 10)]))
-    one = Literal("1", INTEGER_TYPE)
     lbound = BinaryOperation.create(BinaryOperation.Operator.LBOUND,
-                                    Reference(sym), one)
+                                    Reference(sym), _ONE.copy())
     ubound = BinaryOperation.create(BinaryOperation.Operator.UBOUND,
-                                    Reference(sym), one.copy())
+                                    Reference(sym), _ONE.copy())
     array = ArrayReference.create(sym, [Range.create(lbound, ubound)])
     lbnd = array.get_lbound_expression(0)
     assert lbnd.value == "3"
 
 
+def test_aref_to_aos_ubound_expression():
+    '''
+    Test the get_ubound_expression method for an ArrayReference to an array
+    of structures.
+
+    '''
+    sgrid_type = StructureType.create(
+        [("ID", INTEGER_TYPE, Symbol.Visibility.PUBLIC)])
+    sgrid_type_sym = DataTypeSymbol("subgrid_type", sgrid_type)
+    sym = DataSymbol("subgrids", ArrayType(sgrid_type_sym, [(3, 10)]))
+    lbound = BinaryOperation.create(BinaryOperation.Operator.LBOUND,
+                                    Reference(sym), _ONE.copy())
+    ubound = BinaryOperation.create(BinaryOperation.Operator.UBOUND,
+                                    Reference(sym), _ONE.copy())
+    array = ArrayReference.create(sym, [Range.create(lbound, ubound)])
+    ubnd = array.get_ubound_expression(0)
+    assert ubnd.value == "10"
+
+
 def test_member_get_lbound_expression(fortran_writer):
     '''
     Tests for the get_lbound_expression() method when used with a
     sub-class of Member.
 
     '''
-    one = Literal("1", INTEGER_TYPE)
-    two = Literal("2", INTEGER_TYPE)
     # First, test when we don't have type information.
     grid_type = DataTypeSymbol("grid_type", DeferredType())
     sym = DataSymbol("grid_var", grid_type)
-    ref = StructureReference.create(sym, [("data", [one.copy()])])
+    ref = StructureReference.create(sym, [("DATA", [_ONE.copy()])])
```

Comment:
Is this move to uppercase to test the bug you found? Please add a comment.

Reply:
Added